Fix contract call acceptance tests ran via helm test

### Description

Contract call acceptance tests fail when ran via helm test. This is because internally within the cluster we don't have a way to route `/api/v1/contracts/call` to the web3 pods without using a separate URL in the test.

### Steps to reproduce

`./gradlew :test:acceptance --info -Dcucumber.filter.tags=@acceptance`

### Additional context

```
    Caused by: org.springframework.web.reactive.function.client.WebClientResponseException$NotFound: 404 Not Found from POST http://mirror-rest/api/v1/contracts/call
    	at org.springframework.web.reactive.function.client.WebClientResponseException.create(WebClientResponseException.java:223)
    	Suppressed: The stacktrace has been enhanced by Reactor, refer to additional information below: 
    Error has been observed at the following site(s):
    	*__checkpoint ⇢ 404 from POST http://mirror-rest/api/v1/contracts/call [DefaultWebClient]
    Original Stack Trace:
```

### Hedera network

other

### Version

0.74.0-rc2

### Operating system

None

--- a/charts/hedera-mirror/values.yaml
+++ b/charts/hedera-mirror/values.yaml
@@ -376,7 +376,9 @@ test:
           acceptance:
             mirrorNodeAddress: "{{ .Release.Name }}-grpc:5600"
             rest:
-              baseUrl: "http://{{ .Release.Name }}-rest/api/v1"
+              baseUrl: "http://{{ .Release.Name }}-rest"
+            web3:
+              baseUrl: "http://{{ .Release.Name }}-web3"
   cucumberTags: "@acceptance"
   enabled: false
   git:

--- a/hedera-mirror-test/README.md
+++ b/hedera-mirror-test/README.md
@@ -57,7 +57,7 @@ under `hedera.mirror.test.acceptance` include:
 - `operatorKey` - Account private key to be used for signing transaction and client identification. Please do not check
   in.
 - `rest`
-  - `baseUrl` - The host URL to the mirror node. For example, https://testnet.mirrornode.hedera.com/api/v1
+  - `baseUrl` - The host URL to the mirror node. For example, https://testnet.mirrornode.hedera.com
   - `delay` - The time to wait in between failed REST API calls.
   - `maxAttempts` - The maximum number of attempts when calling a REST API endpoint and receiving a 404.
   - `maxBackoff` - The maximum backoff duration the mirror grpc subscriber will wait between attempts.
@@ -70,6 +70,8 @@ under `hedera.mirror.test.acceptance` include:
   - `maxAttempts` - The maximum number of times the sdk should try to submit a transaction to the network.
 - `startupTimeout` - How long the startup probe should wait for the network as a whole to be healthy before failing the
   tests
+- `web3`
+  - `baseUrl` - The endpoint associated with the web3 API.
 - `webclient`
   - `connectionTimeout` - The timeout duration to wait to establish a connection with the server
   - `readTimeout` - The timeout duration to wait for data to be read.

--- a/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java
+++ b/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/client/MirrorNodeClient.java
@@ -30,6 +30,7 @@ import java.util.concurrent.TimeUnit;
 import javax.inject.Named;
 import lombok.NonNull;
 import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang3.StringUtils;
 import org.awaitility.Durations;
 import org.springframework.http.MediaType;
 import org.springframework.web.reactive.function.client.WebClient;
@@ -42,6 +43,7 @@ import com.hedera.hashgraph.sdk.SubscriptionHandle;
 import com.hedera.hashgraph.sdk.TokenId;
 import com.hedera.hashgraph.sdk.TopicMessageQuery;
 import com.hedera.mirror.test.e2e.acceptance.config.AcceptanceTestProperties;
+import com.hedera.mirror.test.e2e.acceptance.config.Web3Properties;
 import com.hedera.mirror.test.e2e.acceptance.props.ContractCallRequest;
 import com.hedera.mirror.test.e2e.acceptance.props.MirrorNetworkNode;
 import com.hedera.mirror.test.e2e.acceptance.props.MirrorNetworkNodes;
@@ -66,12 +68,16 @@ public class MirrorNodeClient {
 
     private final AcceptanceTestProperties acceptanceTestProperties;
     private final WebClient webClient;
+    private final WebClient web3Client;
     private final RetryBackoffSpec retrySpec;
 
     public MirrorNodeClient(AcceptanceTestProperties acceptanceTestProperties,
-                            WebClient webClient) {
+                            WebClient webClient, Web3Properties web3Properties) {
         this.acceptanceTestProperties = acceptanceTestProperties;
         this.webClient = webClient;
+        this.web3Client = StringUtils.isBlank(web3Properties.getBaseUrl()) ? webClient : webClient.mutate()
+                .baseUrl(web3Properties.getBaseUrl())
+                .build();
         var properties = acceptanceTestProperties.getRestPollingProperties();
         retrySpec = Retry.backoff(properties.getMaxAttempts(), properties.getMinBackoff())
                 .maxBackoff(properties.getMaxBackoff())
@@ -260,7 +266,7 @@ public class MirrorNodeClient {
     }
 
     private <T> T callPostRestEndpoint(String uri, Class<T> classType, ContractCallRequest contractCallRequest) {
-        return webClient.post()
+        return web3Client.post()
                 .uri(uri)
                 .body(Mono.just(contractCallRequest), ContractCallRequest.class)
                 .accept(MediaType.APPLICATION_JSON)

--- a/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/RestPollingProperties.java
+++ b/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/RestPollingProperties.java
@@ -30,14 +30,18 @@ import javax.validation.constraints.NotNull;
 import lombok.Data;
 import org.hibernate.validator.constraints.time.DurationMin;
 import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Component;
 import org.springframework.validation.annotation.Validated;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
 
 @Component
 @ConfigurationProperties(prefix = "hedera.mirror.test.acceptance.rest")
 @Data
 @Validated
 public class RestPollingProperties {
+
+    public static final String URL_SUFFIX = "/api/v1";
 
     @NotBlank
     private String baseUrl;
@@ -58,7 +62,18 @@ public class RestPollingProperties {
     private Set<Class> retryableExceptions = Set.of(Exception.class);
 
     public boolean shouldRetry(Throwable t) {
+        // Don't retry negative test cases
+        if (t instanceof WebClientResponseException wcre && wcre.getStatusCode() == HttpStatus.BAD_REQUEST) {
+            return false;
+        }
         return retryableExceptions.stream()
                 .anyMatch(ex -> ex.isInstance(t) || ex.isInstance(Throwables.getRootCause(t)));
+    }
+
+    public String getBaseUrl() {
+        if (baseUrl != null && !baseUrl.endsWith(URL_SUFFIX)) {
+            return baseUrl + URL_SUFFIX;
+        }
+        return baseUrl;
     }
 }

--- a/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/Web3Properties.java
+++ b/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/config/Web3Properties.java
@@ -1,0 +1,44 @@
+package com.hedera.mirror.test.e2e.acceptance.config;
+
+/*-
+ * ‌
+ * Hedera Mirror Node
+ * ​
+ * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
+ * ​
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import static com.hedera.mirror.test.e2e.acceptance.config.RestPollingProperties.URL_SUFFIX;
+
+import javax.inject.Named;
+import lombok.Data;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.validation.annotation.Validated;
+
+@ConfigurationProperties(prefix = "hedera.mirror.test.acceptance.web3")
+@Data
+@Named
+@Validated
+public class Web3Properties {
+
+    private String baseUrl;
+
+    public String getBaseUrl() {
+        if (baseUrl != null && !baseUrl.endsWith(URL_SUFFIX)) {
+            return baseUrl + URL_SUFFIX;
+        }
+        return baseUrl;
+    }
+}

--- a/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java
+++ b/hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/ContractFeature.java
@@ -43,6 +43,7 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.util.ResourceUtils;
 import org.springframework.util.StringUtils;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
 
 import com.hedera.hashgraph.sdk.ContractFunctionParameters;
 import com.hedera.hashgraph.sdk.ContractId;
@@ -54,7 +55,6 @@ import com.hedera.mirror.test.e2e.acceptance.client.MirrorNodeClient;
 import com.hedera.mirror.test.e2e.acceptance.props.CompiledSolidityArtifact;
 import com.hedera.mirror.test.e2e.acceptance.props.MirrorContractResult;
 import com.hedera.mirror.test.e2e.acceptance.props.MirrorTransaction;
-import com.hedera.mirror.test.e2e.acceptance.response.ContractCallResponse;
 import com.hedera.mirror.test.e2e.acceptance.response.MirrorContractResponse;
 import com.hedera.mirror.test.e2e.acceptance.response.MirrorContractResultResponse;
 import com.hedera.mirror.test.e2e.acceptance.util.FeatureInputHandler;
@@ -144,20 +144,24 @@ public class ContractFeature extends AbstractFeature {
 
     @Then("I call the contract via the mirror node REST API")
     public void restContractCall() {
-        ContractCallResponse getAccountBalanceResponse = mirrorClient.contractsCall(GET_ACCOUNT_BALANCE_SELECTOR, contractId.toSolidityAddress(), contractClient.getClientAddress());
+        var from = contractClient.getClientAddress();
+        var to = contractId.toSolidityAddress();
+
+        var getAccountBalanceResponse = mirrorClient.contractsCall(GET_ACCOUNT_BALANCE_SELECTOR, to, from);
         assertThat(convertContractCallResponseToNum(getAccountBalanceResponse)).isEqualTo(BigInteger.valueOf(1000L));
 
-        ContractCallResponse getSenderResponse = mirrorClient.contractsCall(GET_SENDER_SELECTOR, contractId.toSolidityAddress(), contractClient.getClientAddress());
-        assertThat(convertContractCallResponseToAddress(getSenderResponse)).isEqualTo(contractClient.getClientAddress());
+        var getSenderResponse = mirrorClient.contractsCall(GET_SENDER_SELECTOR, to, from);
+        assertThat(convertContractCallResponseToAddress(getSenderResponse)).isEqualTo(from);
 
-        ContractCallResponse multiplySimpleNumbersResponse = mirrorClient.contractsCall(MULTIPLY_SIMPLE_NUMBERS_SELECTOR, contractId.toSolidityAddress(), contractClient.getClientAddress());
+        var multiplySimpleNumbersResponse = mirrorClient.contractsCall(MULTIPLY_SIMPLE_NUMBERS_SELECTOR, to, from);
         assertThat(convertContractCallResponseToNum(multiplySimpleNumbersResponse)).isEqualTo(BigInteger.valueOf(4L));
 
-        ContractCallResponse identifierResponse = mirrorClient.contractsCall(IDENTIFIER_SELECTOR, contractId.toSolidityAddress(), contractClient.getClientAddress());
+        var identifierResponse = mirrorClient.contractsCall(IDENTIFIER_SELECTOR, to, from);
         assertThat(convertContractCallResponseToSelector(identifierResponse)).isEqualTo(IDENTIFIER_SELECTOR);
 
-        assertThatThrownBy(() -> mirrorClient.contractsCall(WRONG_SELECTOR, contractId.toSolidityAddress(), contractClient.getClientAddress())).getCause()
-                .isInstanceOf(Exception.class).hasMessageContaining("400 Bad Request from POST");
+        assertThatThrownBy(() -> mirrorClient.contractsCall(WRONG_SELECTOR, to, from))
+                .isInstanceOf(WebClientResponseException.class)
+                .hasMessageContaining("400 Bad Request from POST");
     }
 
     @Then("the mirror node REST API should verify the deleted contract entity")

--- a/hedera-mirror-test/src/test/resources/application.yml
+++ b/hedera-mirror-test/src/test/resources/application.yml
@@ -4,7 +4,7 @@ hedera:
       acceptance:
         messageTimeout: 20s
         # grpc endpoint
-        mirrorNodeAddress: hcs.testnet.mirrornode.hedera.com:5600
+        mirrorNodeAddress: testnet.mirrornode.hedera.com:443
         network: testnet
         rest:
-          baseUrl: http://testnet.mirrornode.hedera.com/api/v1
+          baseUrl: https://testnet.mirrornode.hedera.com

